### PR TITLE
Partially reversible lithium plating

### PR DESCRIPTION
**Pull request by DrSOKane**
# Description

The partially reversible lithium plating model proposed by O'Kane et al. (2022) is added to PyBaMM.

Fixes # 1807

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [Y] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [Y] No style issues: `$ flake8`
- [N] All tests pass: `$ python run-tests.py --unit`
test_symbol/test_visualize fails
- [Y] The documentation builds: `$ cd docs` and then `$ make clean; make html`

You can run all three at once, using `$ python run-tests.py --quick`.

## Further checks:

- [Y] Code is commented, particularly in hard-to-understand areas
- [Y] Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: pybamm/input/parameters/lithium_ion/lithium_platings/okane2022_Li_plating/parameters.csv [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+Lithium metal partial molar volume [m3.mol-1],1.30E-05,Yang2017,
+Lithium plating kinetic rate constant [m.s-1],1E-9,,
+Exchange-current density for plating [A.m-2],[function]plating_exchange_current_density_OKane2020,,
+Exchange-current density for stripping [A.m-2],[function]stripping_exchange_current_density_OKane2020,,
+Initial plated lithium concentration [mol.m-3],0.00E+00,,
+Typical plated lithium concentration [mol.m-3],1000,1 molar,
+Lithium stripping transfer coefficient,0.35,,
+Lithium plating transfer coefficient,0.65,,
+Dead lithium decay constant [s-1],1E-6,,
+Dead lithium decay rate [s-1],[function]SEI_limited_dead_lithium_OKane2022,,
```
Comment (tinosulzer) [inline pybamm/input/parameters/lithium_ion/lithium_platings/okane2022_Li_plating/parameters.csv:14]: So this parameter set is almost a superset of O'Kane 2020, except the rate constant changes from 1e-10 to 1e-9?

File: pybamm/models/submodels/interface/lithium_plating/plating.py [hunk 2]
```diff
@@ -59,20 +73,26 @@ def get_coupled_variables(self, variables):
...
+        # phi_ref is part of the de-dimensionalization used in PyBaMM
         phi_ref = param.U_n_ref / param.potential_scale
 
         eta_stripping = delta_phi + phi_ref + eta_sei
         eta_plating = -eta_stripping
-        prefactor = 1 / (2 * (1 + self.param.Theta * T))
+        prefactor = 1 / (1 + self.param.Theta * T)
+        # NEW: transfer coefficients can be set by the user
+        alpha_stripping = self.param.alpha_stripping
+        alpha_plating = self.param.alpha_stripping
```
Comment (tinosulzer) [inline pybamm/models/submodels/interface/lithium_plating/plating.py:84]: typo

File: pybamm/models/submodels/interface/lithium_plating/plating.py [hunk 3]
```diff
@@ -85,26 +105,42 @@ def get_coupled_variables(self, variables):
...
 
         Gamma_plating = self.param.Gamma_plating
+        # In the partially reversible plating model, coupling term turns reversible
+        # lithium into dead lithium. In other plating models, it is zero.
+        if self.options["lithium plating"] == "partially reversible":
+            dead_lithium_decay_rate = self.param.dead_lithium_decay_rate(L_sei)
+            coupling_term = dead_lithium_decay_rate * c_plated_Li
+        else:
+            zero = pybamm.Scalar(0)
+            coupling_term = zero
```
Comment (tinosulzer) [inline pybamm/models/submodels/interface/lithium_plating/plating.py:129]: why not just do `coupling_term = pybamm.Scalar(0)` ? This isn't used anywhere else

File: pybamm/parameters/lithium_ion_parameters.py [hunk 4]
```diff
@@ -843,11 +851,15 @@ def _set_dimensionless_parameters(self):
...
-            "Typical plated lithium concentration [mol.m-3]"
-        )
         self.c_plated_Li_0 = self.c_plated_Li_0_dim / self.c_Li_typ
 
+        self.alpha_stripping = pybamm.Parameter(
+            "Lithium stripping transfer coefficient"
+        )
+        self.alpha_plating = pybamm.Parameter(
+            "Lithium plating transfer coefficient"
+        )
```
Comment (tinosulzer) [inline pybamm/parameters/lithium_ion_parameters.py:861]: Should these really be two separate parameters? Shouldn't they add up to 1?

File: tests/unit/test_parameters/test_parameter_sets/test_OKane2020.py [hunk 5]
```diff
@@ -21,6 +21,7 @@ def test_load_params(self):
     def test_functions(self):
         root = pybamm.root_dir()
         param = pybamm.ParameterValues("Chen2020_plating")
+        param.update({"Lithium plating kinetic rate constant [m.s-1]": 1E-10})
```
Comment (tinosulzer) [inline tests/unit/test_parameters/test_parameter_sets/test_OKane2020.py:24]: why is this done? why not update the test value?

**Review by tinosulzer — CHANGES_REQUESTED** [review]
See comments. Also, please add tests for partially reversible plating

File: pybamm/input/parameters/lithium_ion/lithium_platings/okane2022_Li_plating/parameters.csv (hunk 1 above)
Comment (DrSOKane) [reply pybamm/input/parameters/lithium_ion/lithium_platings/okane2022_Li_plating/parameters.csv:14]: Also, the transfer coefficients are different

File: tests/unit/test_parameters/test_parameter_sets/test_OKane2020.py (hunk 5 above)
Comment (DrSOKane) [reply tests/unit/test_parameters/test_parameter_sets/test_OKane2020.py:24]: The two papers used different values for this parameter.

File: pybamm/models/submodels/interface/lithium_plating/plating.py (hunk 2 above)
Comment (DrSOKane) [reply pybamm/models/submodels/interface/lithium_plating/plating.py]: Fixed

File: pybamm/models/submodels/interface/lithium_plating/plating.py (hunk 3 above)
Comment (DrSOKane) [reply pybamm/models/submodels/interface/lithium_plating/plating.py]: Done

**Review by tinosulzer — APPROVED** [review]
This looks great now, thanks @DrSOKane !

File: pybamm/input/parameters/lithium_ion/negative_electrodes/graphite_Chen2020/graphite_LGM50_ocp_Chen2020.py [hunk 6]
```diff
@@ -3,8 +3,9 @@
 
 def graphite_LGM50_ocp_Chen2020(sto):
     """
-    LG M50 graphite open circuit potential as a function of stochiometry, fit taken
-    from [1].
+    LG M50 Graphite open-circuit potential as a function of stochiometry. The fit is
+    taken from [1] with an extra term added by Simon O'Kane to capture behaviour in
+    the high stoichiometry range.
```
[the remark this answers is not in the document]
Comment (DrSOKane) [reply pybamm/input/parameters/lithium_ion/negative_electrodes/graphite_Chen2020/graphite_LGM50_ocp_Chen2020.py:8]: It is out of date. The one in graphite_Chen2020_plating still has this extra term but I removed it from all the others for the sake of consistency with the literature. Shall I make another push that corrects this?

[the remark this answers is not in the document]
Comment (tinosulzer) [reply pybamm/input/parameters/lithium_ion/negative_electrodes/graphite_Chen2020/graphite_LGM50_ocp_Chen2020.py:8]: Yes please

[the remark this answers is not in the document]
Comment (DrSOKane) [reply pybamm/input/parameters/lithium_ion/negative_electrodes/graphite_Chen2020/graphite_LGM50_ocp_Chen2020.py]: Done

**Comment by tinosulzer** [comment]
Forgot to ask for a new line in CHANGELOG, can you do that?

**Comment by DrSOKane** [comment]
I have pushed the new line to my fork, but not sure if that will be pulled in or if I have to make a new request
________________________________

**Pull request merged**